include/dlaf/eigensolver/tridiag_solver/mc.h

```diff
@@ -203,53 +197,49 @@ void offloadDiagonal(Matrix<const T, Device::CPU>& mat_trd, Matrix<T, Device::CP
 /// 2. The overlap between U1' and U2' matches the number of shared columns between Q1' and Q2'
 /// 3. The overlap region is due to deflation via Givens rotations of a column vector from Q1 with a
 ///    column vector of Q2.
-template <class T>
-void TridiagSolver<Backend::MC, Device::CPU, T>::call(Matrix<T, Device::CPU>& mat_trd,
-                                                      Matrix<T, Device::CPU>& d,
-                                                      Matrix<T, Device::CPU>& mat_ev) {
+template <Backend backend, Device device, class T>
+void TridiagSolver<backend, device, T>::call(Matrix<T, device>& mat_trd, Matrix<T, device>& evals,
+                                             Matrix<T, device>& evecs) {
+  // Auxiliary matrix used for the D&C algorithm
+  const matrix::Distribution& distr = evecs.distribution();
+  LocalElementSize vec_size(distr.size().rows(), 1);
+  TileElementSize vec_tile_size(distr.blockSize().rows(), 1);
+  WorkSpace<T, device> ws{Matrix<T, device>(distr),                                // mat1
+                          Matrix<T, device>(distr),                                // mat2
+                          Matrix<T, Device::CPU>(vec_size, vec_tile_size),         // dtmp
+                          Matrix<T, Device::CPU>(vec_size, vec_tile_size),         // z
+                          Matrix<T, Device::CPU>(vec_size, vec_tile_size),         // ztmp
+                          Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),  // i1
+                          Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),  // i2
+                          Matrix<SizeType, Device::CPU>(vec_size, vec_tile_size),  // i3
+                          Matrix<ColType, Device::CPU>(vec_size, vec_tile_size)};  // c
+
+  matrix::MatrixMirror<T, Device::CPU, device> tridiag_h(mat_trd);
+
+  WorkSpaceHostMirror<T, device> ws_h{matrix::MatrixMirror<T, Device::CPU, device>(evals),     // evals
+                                      matrix::MatrixMirror<T, Device::CPU, device>(evecs),     // evecs
+                                      matrix::MatrixMirror<T, Device::CPU, device>(ws.mat1),   // mat1
+                                      matrix::MatrixMirror<T, Device::CPU, device>(ws.mat2)};  // mat2
+
   // Set `mat_ev` to `zero` (needed for Given's rotation to make sure no random values are picked up)
-  matrix::util::set0<Backend::MC, T, Device::CPU>(pika::execution::thread_priority::normal, mat_ev);
+  matrix::util::set0<backend, T, device>(pika::execution::thread_priority::normal, evecs);
+  ws_h.evecs.copySourceToTarget();  // copy from GPU to CPU if evecs is on GPU
```

Comment:
Minor optimization opportunity: `MatrixMirror` currently copies the contents from the source to the target on construction. If I understand this correctly, it would not be necessary in this particular situation because you are anyway operating on the source matrix first, and then copying the contents to the target.

Could you move the construction of `ws_h` to just below the `set0` call here? It looks like that would be sufficient for your use case and avoid the additional copy at the same time.

Reply:
https://github.com/eth-cscs/DLA-Future/pull/617/commits/ff8ad699fd042f36dbc2834b29e12072fd8d8f24